compiler/rustc_typeck/src/check/pat.rs

```diff
@@ -788,11 +788,27 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                             self.tcx.lang_items().range_inclusive_struct(),
                             self.tcx.lang_items().range_to_inclusive_struct(),
                         ];
-                        if const_def_id != None && ranges.contains(&const_def_id) {
+                        if type_def_id != None && ranges.contains(&type_def_id) {
                             let msg = "constants only support matching by type, \
                                 if you meant to match against a range of values, \
                                 consider using a range pattern like `min ..= max` in the match block";
                             e.note(msg);
+
+                            match item_def_id {
+                                Some(def_id) => match self.tcx.hir().get_if_local(def_id) {
+                                    Some(hir::Node::Item(hir::Item {
+                                        kind: hir::ItemKind::Const(_, body_id),
+                                        ..
+                                    })) => {
+                                        e.span_help(
+                                            self.tcx.hir().span(body_id.hir_id),
+                                            "you may want to move the range into the match block",
+                                        );
```

Comment:
Could we instead either use the `body_id` to get the `ItemKind::Expr` of the range or use the `self.tcx.hir().span(body_id.hir_id)` to get its snippet (I would like to verify that the snippet _is_ actually an range expression to avoid nonsensical suggestions in case of something like `const C: RangeInclusive<i32> = panic!();`).

Once we have that, you can do `e.span_suggestion_verbose(ident.span, "you may want to move the range into the `match` block", snippet, Applicability::MachineApplicable);` instead of `span_help`. This has the benefit that the output will show what the resulting code should be and that VSCode will be able to apply this fix with user interaction.

Reply:
Ok just so I understand you want me to actually get the Node::Expr and make sure its a range expression, call the extremely convenient `is_range_literal` fn on it, and only emit this diagnostic when this is true

How do I get the snippet from the Expr?